general/kdtree.hpp

```diff
@@ -0,0 +1,765 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_KDTREE_HPP
+#define MFEM_KDTREE_HPP
+
+#include <vector>
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <cmath>
+
+#include "../linalg/vector.hpp"
+
+namespace mfem
+{
+
+namespace KDTreeNorms
+{
+
+/// Evaluates l1 norm of a vector.
+template <typename Tfloat, int ndim>
+struct Norm_l1
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      if (xx[0]<Tfloat(0.0)) { tm=-xx[0];}
+      else { tm=xx[0];}
+      for (int i=1; i<ndim; i++)
+      {
+         if (xx[i]<Tfloat(0.0)) { tm=tm-xx[i]; }
+         else { tm=tm+xx[i];}
+      }
+      return tm;
+   }
+};
+
+/// Evaluates l2 norm of a vector.
+template<typename Tfloat,int ndim>
+struct Norm_l2
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      tm=xx[0]*xx[0];
+      for (int i=1; i<ndim; i++)
+      {
+         tm=tm+xx[i]*xx[i];
+      }
+      return sqrt(tm);
+   }
+};
+
+/// Finds the max absolute value of a vector.
+template<typename Tfloat,int ndim>
+struct Norm_li
+{
+   Tfloat tm;
+   Tfloat operator() (const Tfloat* xx)
+   {
+      if (xx[0]<Tfloat(0.0)) { tm=-xx[0];}
+      else { tm=xx[0];}
+      for (int i=1; i<ndim; i++)
+      {
+         if (xx[i]<Tfloat(0.0))
+         {
+            if (tm<(-xx[i])) {tm=-xx[i];}
+         }
+         else
+         {
+            if (tm<xx[i]) {tm=xx[i];}
+         }
+      }
+      return tm;
+   }
+};
+
+}
+
+/// Template class for build KDTree with template parameters Tindex
+/// specifying the type utilized for indexing the points, Tfloat
+/// specifying a float type for representing the coordinates of the
+/// points, integer parameter ndim specifying the dimensionality of the
+/// space and template function Tnorm for evaluating the distance
+/// between two points.
+template <typename Tindex, typename Tfloat, size_t ndim=3, typename Tnorm=KDTreeNorms::Norm_l2<Tfloat,ndim> >
```

Comment:
The norm struct/function will define the support domain for mesh-less/free interpolations/approximations. So it will impact the definition of these approximations (representation as tensor product, directional dependence,  local rotation of the space, etc.). One should be able to control these norms.

Reply:
I suggest that the doxygen comments should include why the KDTree is being implemented and what role it plays in MFEM.  For example, consider this text:

The KDTree class implements the standard <a href="https://en.wikipedia.org/wiki/K-d_tree">k-d tree</a> data structure that will be used to transfer a ParGridFunction defined on one MPI communicator to a ParGridFunction/GridFunction defined on different MPI communicator. This can be useful when comparing a solution computed on `m` ranks against a solution computed with `n` or 1 rank(s).

Reply:
Description is added.